### DISCUSSION TITLE
How can I filter results from a dataloader?

### DISCUSSION THREAD

**Discussion by peterbierman** [Netflix/dgs-framework#461] [Q&A] [answered]
As is documented here, https://netflix.github.io/dgs/data-loaders/.
You can use dataloaders to load a list of objects from a service.

I want to filter this list after the result is returned.
So in this example for instance, i just want directors for which the name starts with an "A".

This code calls the dataloader and immediatly returns it.

`@DgsData(parentType = "Movie", field = "director")`
`    public CompletableFuture<Director> director(DataFetchingEnvironment dfe) {`
`        DataLoader<String, Director> dataLoader = dfe.getDataLoader("directors");`
`        String id = dfe.getArgument("directorId");`
`        return dataLoader.load(id);`
`    }`

But I first need to filter the result before I return it, how is this possible?
I tried running the dataloader with a `dispatchAndJoin()` call but this just returns the results that were not fetched previously so calls that were already loaded and cached don't return then.

**Comment by paulbakker** [comment, marked as answer]
You can for example use `thenApply` on the `CompletableFuture`, which is similar to `map` on a stream.

```
return CompletableFuture.supplyAsync(() -> reviewsService.reviewsForShows(new ArrayList<>(keys))).thenApply(r -> {
            Map<Integer, List<Review>> filtered = new HashMap<>();

            r.keySet().forEach(key -> {
                filtered.put(key, r.get(key).stream().filter(review -> review.getStarScore() > 3).collect(Collectors.toList()));
            }) ;

            return filtered;
        });
```

There are probably nicer ways to write the same code, but this gives an idea of what you can do.
